src/Sylius/Bundle/ApiBundle/Serializer/ProductNormalizer.php

```diff
@@ -0,0 +1,57 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\Serializer;
+
+use ApiPlatform\Core\Api\IriConverterInterface;
+use Sylius\Component\Core\Model\ProductInterface;
+use Sylius\Component\Product\Resolver\ProductVariantResolverInterface;
+use Symfony\Component\Serializer\Normalizer\ContextAwareNormalizerInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareInterface;
+use Symfony\Component\Serializer\Normalizer\NormalizerAwareTrait;
+use Webmozart\Assert\Assert;
+
+/** @experimental */
+final class ProductNormalizer implements ContextAwareNormalizerInterface, NormalizerAwareInterface
+{
+    use NormalizerAwareTrait;
+
+    private const ALREADY_CALLED = 'product_normalizer_already_called';
+
+    /** @var ProductVariantResolverInterface */
+    private $defaultProductVariantResolver;
+
+    /** @var IriConverterInterface */
+    private $iriConverter;
+
+    public function __construct(
+        ProductVariantResolverInterface $defaultProductVariantResolver,
+        IriConverterInterface $iriConverter
+    ) {
+        $this->defaultProductVariantResolver = $defaultProductVariantResolver;
+        $this->iriConverter = $iriConverter;
+    }
+
+    public function normalize($object, $format = null, array $context = [])
+    {
+        Assert::isInstanceOf($object, ProductInterface::class);
+
+        $context[self::ALREADY_CALLED] = true;
```

Comment:
Perhaps we should add assert of early return in case this class is already called indeed?

Reply:
IMO there is no need, because this check is in `supportsNormalization` method, but I can add it to be in 100% secured